src/ui/controllers/searchbariconcontroller.js

```diff
@@ -0,0 +1,83 @@
+import SearchStates from '../../core/storage/searchstates';
+import StorageKeys from '../../core/storage/storagekeys';
+import DOM from '../dom/dom';
+import LoadingIconState from '../statemachine/loadingiconstate';
+import SearchIconState from '../statemachine/searchiconstate';
+import DefaultIconState from '../statemachine/defaulticonstate';
+import { StateMachine } from '../statemachine/statemachine';
+
+/**
+ * Responsible for controlling the search bar icon behavior
+ */
+export default class SearchBarIconController {
+  constructor (config) {
+    this.iconIsFrozen = false;
+    this.inputEl = config.inputEl;
+    this.searchBarContainer = config.searchBarContainer;
+    this.onMouseUp = this.oneTimeMouseUpListener.bind(this);
+
+    const IconState = {
```

Comment:
nit: To me this reads like it's an enum where the values are strings, and that throws me off a little since the values are class instances. Is it possible to make this a regular enum? Each state class could have the enum set as its own id.  Each instance of the state could be passed to the state machine. Then when the state machine transitions it could switch to the correct state instance

Reply:
I see what you mean. I will have the enum store the id and instantiate the states separately in the update.